### PR TITLE
refactor(skills): split SkillService into cohesive modules

### PR DESCRIPTION
**Pull request by DeJeune**
> ### Branch strategy
>
> - Active development targets `main`.
> - v1 maintenance targets `v1`; forward-port fixes to `main` separately when needed.

### What this PR does

Before this PR:

`SkillService.ts` was 1840 lines covering the skill catalog, four marketplace fetchers, git plumbing, ZIP handling, path containment checks, the Claude config mirror, reconcile and builtin sync. That size is why #18549 originally landed 120 lines of Claude Agent SDK runtime logic inside it: for anything skill-shaped, this file was the default destination. `isOutsidePath` was heading the same way — it is a path predicate, and this branch routes it to `@main/utils/file`, which already owns those, rather than adding a fourth containment check to the repo.

After this PR:

The stateless parts are extracted into three modules and `SkillService` drops to ~1150 lines:

- `skillRemoteSource.ts` — the `claude-plugins` / `skills.sh` / `clawhub` / `github` fetchers plus `ls-remote` ref resolution, sparse commit fetch, clone and default-branch probing. Reduced to one contract: fill a temp directory and report which directory in it is the skill.
- `skillArchive.ts` — ZIP validation and extraction, the size/file ceilings applied to both archives and checkouts, and every containment check (`resolveSkillDirectory`, `validateRepositorySkillDirectory`, symlink-component rejection) that decides which directory of an untrusted tree gets installed.
- `skillPaths.ts` — folder-name sanitizing and case folding, temp workspace creation, recursive removal, file-tree building.

`isOutsidePath` moves to `@main/utils/file` (`./path`, alongside `isPathInside` / `isSameOrInside`) instead of staying private to the skills modules.

`SkillService` keeps the catalog surface, the four lock-holding writers, `installSkillDirLocked`, ZIP/directory installs, local and system discovery, reconcile, the mirror and builtin sync.

Fixes # — no linked GitHub issue.

### Why we need it and why it was done in this way

The split line is the mutation lock, not file size. `uninstall`, `installSkillDir`, `reconcileSkills` and `syncBuiltinSkill` serialize every library write through one `Mutex`, so any module that could reach the catalog would have to reach the lock with it. Everything extracted here is stateless and touches neither, which is what makes the move mechanical rather than a redesign.

The following tradeoffs were made:

- `install()` now owns the temp workspace it disposes of, instead of each fetcher creating and cleaning up its own. The fetchers receive a lazy `openTempDir()` and call it at exactly the point they previously created the directory, so a malformed identifier is still rejected before anything is written to disk, and the temp-dir prefix comes from the matched source key rather than raw user input.
- The Claude config mirror stays in `SkillService`. `linkMirror` reads catalog rows, builtin content hashes and descriptor state, so extracting it would mean injecting four callbacks — an abstraction, not a move. It is flagged here rather than half-done.
- `isOutsidePath` stays a separate export rather than folding into `isSameOrInside`. It takes the relative path the caller already computed, treats `''` as inside, matches `..` only as a whole segment, and skips the platform case folding `isPathInside` applies. The segment difference is load-bearing here: `path.relative('/repo', '/repo/..archive')` is `'..archive'`, which `isPathInside`'s `startsWith('..')` reads as an escape, and an in-repository directory named `..archive` is covered by an existing install test. That `isPathInside` false positive is left alone — it is shared with `ipc.ts`, `fileOperations.ts` and a third hand-rolled copy in `core/security/validateSender.ts`, so unifying them is its own change.
- Three tests were deleted: each spied on a private `installFromX` and asserted `install()` called it with its own argument. They cannot fail for a real reason, and every route stays covered by the tests that exercise it end to end.

The following alternatives were considered:

- Split by file size into arbitrary chunks: rejected because it would put catalog writes on both sides of the lock boundary.
- Keep everything and only extract the mirror: rejected because the mirror is the one part genuinely welded to catalog state, while the fetchers and archive handling are not.
- Fold the extraction into #18549: rejected because a 40-line semantic fix buried in a 1200-line move is not reviewable, and a dropped `runExclusive` during the move would need its own bisect.

Links to places where the discussion took place: review of #18549.

### Breaking changes

None. No public API, no behavior, no persisted data changes.

### Special notes for your reviewer

Verification that this is a move and not a rewrite:

- `SkillService.test.ts` assertions are unchanged. What changed is the receiver: `skillService['extractZip'](...)` became `skillArchive.extractZip(...)`, and private-method spies became module spies wrapping the real implementations (`vi.mock(..., factory)` + `vi.fn(actual.fn)`), so every install test still runs the real path with one step stubbed.
- The module namespace import is deliberate: the test file already defines a local `createTempDir` helper that shadowed a named import.
- A per-test `mockReset()` was added because module spies, unlike per-test `spyOn`, persist across tests.
- `isOutsidePath` gained four direct assertions in `utils/file/__tests__/path.test.ts`; the `..archive` one exists specifically to fail if someone later collapses it into `isPathInside`.
- Ran: `pnpm vitest run src/main/ai src/main/ipc src/main/utils` — 4685 tests, 343 files, all passing.
- Ran: `pnpm format`, `npx tsgo --noEmit -p tsconfig.node.json`, `npx oxlint src/main/ai/skills` (0 warnings, 0 errors), `npx eslint --max-warnings 0 src/main/ai/skills`.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Branch: This PR targets the correct branch — `main` for active development, `v1` for v1 maintenance fixes
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
NONE
```

### CODE REVIEW

**Review by 404-Page-Found — CHANGES_REQUESTED** [review]
## Review — `refactor(skills): split SkillService into cohesive modules`

**Overall assessment:** Well-structured mechanical extraction. The PR description is thorough and the split line (stateless vs lock-holding) is sound. Three issues warrant attention before merge.

### Must fix

| # | Severity | File | Finding |
|---|----------|------|---------|
| 1 | **Major** | `skillRemoteSource.ts` | **`getProxyEnvironment` dropped from `runGit`** — the original `SkillService.runGit()` merged `getProxyEnvironment(process.env)` into the subprocess env, which was lost during extraction. All git operations (clone, fetch, ls-remote, ls-tree) no longer respect user proxy settings. This is a silent behavioral regression for users behind corporate proxies or VPNs, and contradicts the "no behavior changes" claim in the PR description. **Fix:** add `...getProxyEnvironment(process.env)` back to the env object in `runGit`, and add the import. |
| 2 | **Major** | `skillRemoteSource.ts` | **`cloneRepository` bypasses hardened `runGit`** — it calls `executeCommand` directly without the hardened env (`GIT_TERMINAL_PROMPT=0`, `GIT_LFS_SKIP_SMUDGE=1`, `GIT_ASKPASS=''`, `GCM_INTERACTIVE=never`) or the 2-minute timeout. A large repo with LFS objects will block on smudge; a repo behind HTTP basic auth will hang on a credential prompt. **Fix:** route `cloneRepository` and `resolveDefaultBranch` through `runGit`. |
| 3 | **Major** | `skillRemoteSource.ts` | **`assertUniqueSkillPath` truncates via `slice(0, -1)` when `indexOf` returns -1** — when the entry path exactly equals the directory path (no trailing `/`), `indexOf('/', directoryPath.length)` returns -1, and `line.slice(0, -1)` silently truncates the last character, polluting the collision set. **Fix:** add `if (idx === -1) continue` guard. |

### Should fix

| # | Severity | File | Finding |
|---|----------|------|---------|
| 4 | **Minor** | `skillRemoteSource.ts` | `resolveDefaultBranch` catches all errors from `ls-remote` and returns `null` — this swallows unexpected failures (git binary not found, permission denied) and the user gets a generic clone error. Consider `logger.debug` in the catch. |
| 5 | **Minor** | `skillRemoteSource.ts` | `reportInstall` `.catch(...)` logs `err.message` but discards the stack trace. Log the full error object for diagnosability. |

### Acceptable / nits

| # | Severity | File | Finding |
|---|----------|------|---------|
| 6 | Nit | `skillArchive.ts` | `validateZipFile` name overpromises — it's a stat + extension check, not ZIP integrity validation. Fine in practice (extraction catches corruption). |
| 7 | Nit | `skillArchive.ts` | Size/count ceilings trust zip metadata headers, not actual bytes written (zip bomb vector). Acceptable for a desktop app where the user is the attacker. |
| 8 | Nit | `skillRemoteSource.ts` | `cloneRepository` default branch probing is a behavioral improvement, not a mechanical extraction — worth calling out explicitly in the PR description. |

### What's good

- **Mutex/lock logic** preserved exactly — all four `runExclusive` call sites intact
- **Import wiring** is clean — no orphaned imports, all extracted functions properly re-imported
- **`isOutsidePath` extraction** is correct, well-tested, and properly handles the `..archive` edge case
- **Test changes** are well-structured: private-method spies correctly converted to module spies, `mockReset()` applied properly, deleted tests were pure behavior-pinning
- **`install()` temp workspace lifecycle** correctly centralized with the lazy `openTempDir()` pattern
- **Module boundaries** are clean — everything extracted is stateless and lock-free, everything kept touches catalog state

The three major items (#1-3) are straightforward to fix. The proxy regression (#1) is the most impactful since it's a silent behavioral change. The `cloneRepository` bypass (#2) and the `slice` bug (#3) are correctness issues.

**Comment by 404-Page-Found** [comment]
### Inline findings on `skillRemoteSource.ts`

Three major issues worth highlighting with exact locations:

---

**1. `getProxyEnvironment` dropped from `runGit` (line ~380)**

The original `SkillService.runGit()` merged `getProxyEnvironment(process.env)` into the subprocess env, which was lost during extraction. All git operations (clone, fetch, ls-remote, ls-tree) no longer respect user proxy settings — a silent behavioral regression for users behind corporate proxies or VPNs.

```typescript
// Current (broken)
env: { ...env, GIT_TERMINAL_PROMPT: '0', ... }

// Should be
env: { ...env, ...getProxyEnvironment(process.env), GIT_TERMINAL_PROMPT: '0', ... }
```

---

**2. `cloneRepository` bypasses hardened `runGit` (line ~386)**

Calls `executeCommand` directly without the hardened env or the `GIT_COMMAND_TIMEOUT_MS` timeout. A large repo with LFS objects will block on smudge; a repo behind HTTP basic auth will hang on a credential prompt.

---

**3. `assertUniqueSkillPath` truncates via `slice(0, -1)` when `indexOf` returns -1 (line ~360)**

```typescript
const idx = line.indexOf('/', directoryPath.length)
// When idx === -1, line.slice(0, -1) truncates the last character
collisions.add(line.slice(0, idx))
```

Fix: add `if (idx === -1) continue` before the slice.

**Comment by 404-Page-Found** [comment]
Correction — here are the exact line numbers in `src/main/ai/skills/skillRemoteSource.ts`:

**Line 311** — `assertUniqueSkillPath` truncation bug:
```typescript
collisions.add(line.slice(0, line.indexOf('/', directoryPath.length)))
```

**Line 327** — `runGit` missing `getProxyEnvironment`:
```typescript
env: { ...env, GIT_TERMINAL_PROMPT: '0', GIT_LFS_SKIP_SMUDGE: '1', GIT_ASKPASS: '', GCM_INTERACTIVE: 'never' }
```

**Lines 336, 341, 343** — `cloneRepository` calls `executeCommand` directly, bypassing `runGit`'s hardened env and timeout.

File: src/main/ai/skills/skillRemoteSource.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    )
+  }
+}
+
+async function runGit(gitCommand: string, args: string[]): Promise<string> {
+  const env = await getShellEnv()
+  return executeCommand(gitCommand, args, {
+    capture: true,
+    timeout: GIT_COMMAND_TIMEOUT_MS,
+    env: { ...env, GIT_TERMINAL_PROMPT: '0', GIT_LFS_SKIP_SMUDGE: '1', GIT_ASKPASS: '', GCM_INTERACTIVE: 'never' }
```
Comment (404-Page-Found) [inline src/main/ai/skills/skillRemoteSource.ts]: **Major — `getProxyEnvironment` dropped from `runGit`**

The original `SkillService.runGit()` merged `getProxyEnvironment(process.env)` into the subprocess env, which was lost during extraction. All git operations (clone, fetch, ls-remote, ls-tree) no longer respect user proxy settings — a silent behavioral regression for users behind corporate proxies or VPNs.

**Fix:** import `getProxyEnvironment` and add `...getProxyEnvironment(process.env)` to the env object.

File: src/main/ai/skills/skillRemoteSource.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+  })
+}
+
+async function cloneRepository(repoUrl: string, destDir: string): Promise<void> {
+  const gitCommand = (await findExecutableInEnv('git')) ?? 'git'
+
+  const branch = await resolveDefaultBranch(gitCommand, repoUrl)
+  if (branch) {
+    await executeCommand(gitCommand, ['clone', '--depth', '1', '--branch', branch, '--', repoUrl, destDir])
```
Comment (404-Page-Found) [inline src/main/ai/skills/skillRemoteSource.ts]: **Major — `cloneRepository` bypasses hardened `runGit`**

Calls `executeCommand` directly without the hardened env (`GIT_TERMINAL_PROMPT=0`, `GIT_LFS_SKIP_SMUDGE=1`, `GIT_ASKPASS=''`, `GCM_INTERACTIVE=never`) or the `GIT_COMMAND_TIMEOUT_MS` timeout. A large repo with LFS objects will block on smudge; a repo behind HTTP basic auth will hang on a credential prompt.

**Fix:** route `cloneRepository` (and `resolveDefaultBranch`) through `runGit`, or at minimum pass the same hardened env and timeout.

File: src/main/ai/skills/skillRemoteSource.ts [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+  const gitCommand = (await findExecutableInEnv('git')) ?? 'git'
+  const output = await runGit(gitCommand, ['-C', repoDir, 'ls-tree', '-r', '--name-only', 'FETCH_HEAD'])
+  const foldKey = (value: string) => value.normalize('NFC').toLowerCase()
+  const wanted = foldKey(`${directoryPath}/`)
+
+  const collisions = new Set<string>()
+  for (const entry of output.split('\n')) {
+    const line = entry.trim()
+    if (!line || !foldKey(line).startsWith(wanted)) continue
+    collisions.add(line.slice(0, line.indexOf('/', directoryPath.length)))
```
Comment (404-Page-Found) [inline src/main/ai/skills/skillRemoteSource.ts]: **Major — `assertUniqueSkillPath` truncates via `slice(0, -1)` when `indexOf` returns -1**

When `line` exactly equals `directoryPath` (no trailing `/`), `indexOf('/', directoryPath.length)` returns -1, and `line.slice(0, -1)` silently truncates the last character, polluting the collision set.

**Fix:** add `if (idx === -1) continue` guard:
```typescript
const idx = line.indexOf('/', directoryPath.length)
if (idx === -1) continue
collisions.add(line.slice(0, idx))
```

**Review by 404-Page-Found — CHANGES_REQUESTED** [review]
Follow-up inline findings on `skillRemoteSource.ts` — three major issues with exact line references (see top-level review for full context).

File: src/main/ai/skills/skillRemoteSource.ts (hunk 1 above)
Comment (DeJeune) [reply src/main/ai/skills/skillRemoteSource.ts]: Fixed in e4a4bad375 — but the cause is different from "lost during extraction", and it matters for how this PR gets integrated.

This branch forked at `33213cbdda`, before `1255f87b71 fix(skills): stop online skill install from hanging on git clone (#18677)` landed on `main` (2026-08-17). `getProxyEnvironment` was never in the `runGit` I moved — the copy in `skillRemoteSource.ts` is that fix's *before* state, faithfully moved. So this is not an extraction defect; it is this branch being stale enough that merging it would have **reverted #18677** for every marketplace install.

`runGit` now layers `getProxyEnvironment(process.env)` over the login shell env, matching `main`. #18677's proxy regression test is ported alongside it (`hands git the proxy Cherry is configured with…`), adapted to spy the moved module rather than the private method.

File: src/main/ai/skills/skillRemoteSource.ts (hunk 2 above)
Comment (DeJeune) [reply src/main/ai/skills/skillRemoteSource.ts]: Fixed in e4a4bad375, same root cause as the `getProxyEnvironment` thread: this branch predates #18677, so the 15-line `cloneRepository` (ls-remote --symref → clone → retry on `--branch master`) is what SkillService actually had when I moved it, not something the extraction loosened.

`cloneRepository` is now the post-#18677 shape — one bounded `git clone --depth 1` through `runGit`, so it gets `GIT_COMMAND_TIMEOUT_MS` plus the hardened env — and `resolveDefaultBranch` is deleted with the branch resolution it existed for. Two of #18677's regression tests come with it: `bounds a clone and blocks its credential prompts…` asserts every git call carries a timeout and `GIT_TERMINAL_PROMPT=0` / `GIT_ASKPASS=`, and `surfaces a failed clone…` asserts exactly one attempt on failure.

File: src/main/ai/skills/skillRemoteSource.ts (hunk 3 above)
Comment (DeJeune) [reply src/main/ai/skills/skillRemoteSource.ts]: Not reproducing this one — the stated trigger is unreachable, and the line is byte-identical on `main`, so it is not something this PR introduced.

`wanted` carries a trailing slash (`foldKey(`${directoryPath}/`)`), so a `ls-tree` line equal to `directoryPath` fails `startsWith(wanted)` and hits `continue` before the `slice`. Ran it:

```
directoryPath = "skills/demo"

["skills/demo"]                          -> []                                 // the reported case
["skills/demo/SKILL.md"]                 -> ["skills/demo"]
["skills/demo/SKILL.md",
 "skills/Demo/SKILL.md"]                 -> ["skills/demo", "skills/Demo"]     // collision still detected
```

Every line that reaches the `slice` has `directoryPath/` as a prefix, so there is always a `/` at or after `directoryPath.length` and `indexOf` cannot return -1. `ls-tree -r --name-only` also only ever lists blobs, so a bare directory line never appears.

I left it alone rather than adding a defensive guard: the identical code is live on `main`, and this PR is premised on being a pure move. If you want the guard anyway, happy to send it as a one-liner against `main` — that is also where it would actually protect users.

**Review by 404-Page-Found — COMMENTED** [review]
## Review: Split SkillService into Cohesive Modules

**Overall: Approve** — This is a clean, well-motivated refactor. The split line (the mutation lock boundary) is well-chosen: everything extracted is stateless and touches neither the catalog nor the lock. The PR description is exceptional and the commit history tells a clear story.

### What looks correct

- **Import cleanup**: All imports removed from `SkillService.ts` (`findExecutableInEnv`, `deleteDirectoryRecursive`, `findAllSkillDirectories`, `executeCommand`, `ClawhubSkillDetailSchema`, `encodeGithubPath`, `parseGithubSkillUrl`, `resolveRefFromSegments`, `net`, `StreamZip`) are accounted for in the new modules.
- **Constants moved correctly**: `MAX_EXTRACTED_SIZE`/`MAX_FILES_COUNT` → `skillArchive.ts`; `MAX_FOLDER_NAME_LENGTH` → `skillPaths.ts`; `CLAUDE_PLUGINS_API`/`GIT_COMMAND_TIMEOUT_MS` → `skillRemoteSource.ts`.
- **`isOutsidePath` extraction**: Semantics preserved exactly. The `..archive` test case guards against future collapse into `isPathInside`.
- **Temp workspace ownership**: The lazy `openTempDir()` pattern is correct — malformed identifier rejected before anything hits disk.
- **Test coverage preserved**: All critical paths covered. The three deleted tests were behavior-pinning tests per CLAUDE.md rules.
- **Security**: Path traversal blocked, symlink component rejection preserved, ZIP ceilings preserved, git commands bounded and non-interactive.
- **No API contract changes**: All public methods on `SkillService` remain identical.

### Findings (all nits/minor — none block merging)

1. **Nit** — Empty section comment left in `SkillService.ts` (see inline)
2. **Nit** — `readFile` manual path check vs `isOutsidePath` (see inline) — correct, just a note for maintainers
3. **Minor** — `assertUniqueSkillPath` benign `indexOf(-1)` edge case (see inline)
4. **Nit** — `FetchedSkill` export scope (see inline)

**Comment by 404-Page-Found** [comment]
### Detailed findings

**1. Nit — Empty section comment in `SkillService.ts`**

The `// Source-specific install flows` section header (around line 335 in the PR) now sits above an empty section — the fetcher methods it guarded were moved to `skillRemoteSource.ts`. The `// Core install logic` section immediately follows. Harmless but reads as a leftover heading. Consider removing it.

---

**2. Nit — `readFile` manual path check vs `isOutsidePath`**

`readFile` still uses a hand-rolled `startsWith` + equality check rather than the newly extracted `isOutsidePath`:
```typescript
if (!filePath.startsWith(skillRoot + path.sep) && filePath !== skillRoot) return null
```
This is **correct** — `isOutsidePath` takes a relative path and treats `''` (the base itself) as inside, which wouldn't work here since `readFile` needs to explicitly allow reading the root file itself. The manual check is more precise for this use case. A brief comment might help future maintainers understand why the shared utility isn't used here, but this is entirely optional.

---

**3. Minor — `assertUniqueSkillPath` benign `indexOf(-1)` edge case**

In `skillRemoteSource.ts`, `assertUniqueSkillPath` does:
```typescript
collisions.add(line.slice(0, line.indexOf('/', directoryPath.length)))
```
If a bare directory name (no trailing `/` or child file) matched the prefix, `indexOf` would return `-1` and `slice(0, -1)` would trim the last character. In practice `git ls-tree -r` only lists files (not bare directories), so this can never trigger. Not a bug, but a guard would make the intent clearer:
```typescript
const sepIndex = line.indexOf('/', directoryPath.length)
collisions.add(sepIndex === -1 ? line : line.slice(0, sepIndex))
```

---

**4. Nit — `FetchedSkill` export scope**

`skillRemoteSource.ts` exports the `FetchedSkill` interface, but the skills directory has no `index.ts` barrel and the only consumer is `SkillService.ts` in the same directory. Fine in practice — just noting that if a barrel were ever added, this would need consideration.

---

**5. Observation — Base branch**

This PR targets `feat/skill-missing-dep-silent-fallback`, not `main`. Intentional per the branch strategy — just flagging that a follow-up merge to `main` will be needed.

**Review by 404-Page-Found — APPROVED** [review]
